In [2]:
%load_ext sql

In [3]:
%sql mysql+mysqlconnector://root:Sd123258@localhost:3306/hr_report

In [27]:
%%sql

SELECT * FROM hr LIMIT 5;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
5 rows affected.


[('00-0037846', 'Kimmy', 'Walczynski', datetime.date(1991, 6, 4), 32, 'Male', 'Hispanic or Latino', 'Engineering', 'Programmer Analyst I', 'Headquarters', datetime.date(2002, 1, 20), '0000-00-00', 'Cleveland', 'Ohio'),
 ('00-0041533', 'Ignatius', 'Springett', datetime.date(1984, 6, 29), 39, 'Male', 'White', 'Business Development', 'Business Analyst', 'Headquarters', datetime.date(2019, 4, 8), '0000-00-00', 'Cleveland', 'Ohio'),
 ('00-0045747', 'Corbie', 'Bittlestone', datetime.date(1989, 7, 29), 33, 'Male', 'Black or African American', 'Sales', 'Solutions Engineer Manager', 'Headquarters', datetime.date(2010, 10, 12), '0000-00-00', 'Cleveland', 'Ohio'),
 ('00-0055274', 'Baxy', 'Matton', datetime.date(1982, 9, 14), 40, 'Female', 'White', 'Services', 'Service Tech', 'Headquarters', datetime.date(2005, 4, 10), '0000-00-00', 'Cleveland', 'Ohio'),
 ('00-0076100', 'Terrell', 'Suff', datetime.date(1994, 4, 11), 29, 'Female', 'Two or More Races', 'Product Management', 'Business Analyst', 'Remote', datetime.date(2010, 9, 29), '2029-10-29', 'Flint', 'Michigan')]

In [28]:
%%sql
DESCRIBE hr

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
14 rows affected.


[('Custom_ID', b'text', 'YES', '', None, ''),
 ('first_name', b'text', 'YES', '', None, ''),
 ('last_name', b'text', 'YES', '', None, ''),
 ('birthdate', b'date', 'YES', '', None, ''),
 ('Age', b'int', 'YES', '', None, ''),
 ('gender', b'text', 'YES', '', None, ''),
 ('race', b'text', 'YES', '', None, ''),
 ('department', b'text', 'YES', '', None, ''),
 ('jobtitle', b'text', 'YES', '', None, ''),
 ('location', b'text', 'YES', '', None, ''),
 ('hire_date', b'date', 'YES', '', None, ''),
 ('termdate', b'text', 'YES', '', None, ''),
 ('location_city', b'text', 'YES', '', None, ''),
 ('location_state', b'text', 'YES', '', None, '')]

## Questions

1. How is the gender composition distributed among employees within the company?
2. Could you provide an overview of the distribution of race and ethnicity among the company's workforce?
3. Can you share insights regarding the distribution of employee ages across the company?
4. What is the division between employees working at the company's headquarters and those based in remote locations?
5. Could you offer information about the average duration of employment for employees whose tenure has concluded?
6. How do gender proportions differ across various departments and job titles?
7. Can you outline the prevalence of different job titles across the entire company?
8. Which department exhibits the highest rate of turnover among its employees?
9. What is the dispersion of employees across various city and state locations?
10. Can you chart the historical shifts in the company's employee count over time, considering both hires and terminations?

### 1. How is the gender composition distributed among employees within the company? 
    -- The gender distribution is calculated by counting the occurrences of "Male" and "Female" values in the gender column of the HR Database.

In [29]:
%%sql
SELECT gender, COUNT(*) AS count
FROM hr
WHERE age >= 17
GROUP BY gender

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
3 rows affected.


[('Male', 10794), ('Female', 9876), ('Non-Conforming', 577)]

### 2. Could you provide an overview of the distribution of race and ethnicity among the company's workforce?

    -- The race/ethnicity breakdown is obtained by counting the occurrences of different race/ethnicity values in the race column of the HR Database Table.

In [30]:
%%sql
SELECT race, COUNT(*) AS count
FROM hr
WHERE age >= 17
GROUP BY race
ORDER BY count DESC

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
7 rows affected.


[('White', 6057),
 ('Two or More Races', 3508),
 ('Black or African American', 3443),
 ('Asian', 3388),
 ('Hispanic or Latino', 2402),
 ('American Indian or Alaska Native', 1273),
 ('Native Hawaiian or Other Pacific Islander', 1176)]

### 3. Can you share insights regarding the distribution of employee ages across the company?
    -- The age distribution is determined by calculating the range of ages based on birthdate information and counting the occurrences within each age range.


In [31]:
%%sql
SELECT 
  MIN(age) AS youngest,
  MAX(age) AS oldest
FROM hr
WHERE age >= 17

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
1 rows affected.


[(20, 57)]

In [32]:
%%sql
SELECT FLOOR(age/10)*10 AS age_group, COUNT(*) AS count
FROM hr
WHERE age >= 17
GROUP BY FLOOR(age/10)*10;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
4 rows affected.


[(30, 6159), (40, 5897), (20, 5500), (50, 3691)]

### Using class interval to group the age of employee, to know the total number of employee's who falls in a certain age Group

In [36]:
%%sql
SELECT 
  CASE 
    WHEN age >= 17 AND age <= 24 THEN '17-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, 
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 17
GROUP BY age_group
ORDER BY age_group;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
5 rows affected.


[('17-24', 2508),
 ('25-34', 6074),
 ('35-44', 6017),
 ('45-54', 5655),
 ('55-64', 993)]

In [35]:
%%sql
SELECT 
  CASE 
    WHEN age >= 17 AND age <= 24 THEN '17-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, gender,
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 17
GROUP BY age_group, gender
ORDER BY age_group, gender;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
15 rows affected.


[('17-24', 'Female', 1134),
 ('17-24', 'Male', 1313),
 ('17-24', 'Non-Conforming', 61),
 ('25-34', 'Female', 2863),
 ('25-34', 'Male', 3053),
 ('25-34', 'Non-Conforming', 158),
 ('35-44', 'Female', 2708),
 ('35-44', 'Male', 3145),
 ('35-44', 'Non-Conforming', 164),
 ('45-54', 'Female', 2684),
 ('45-54', 'Male', 2809),
 ('45-54', 'Non-Conforming', 162),
 ('55-64', 'Female', 487),
 ('55-64', 'Male', 474),
 ('55-64', 'Non-Conforming', 32)]

### 4. What is the division between employees working at the company's headquarters and those based in remote locations?

    -- The age distribution is determined by calculating the range of ages based on birthdate information and counting the occurrences within each age range.

In [37]:
%%sql
SELECT location, COUNT(*) as count
FROM hr
WHERE age >= 17
GROUP BY location;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
2 rows affected.


[('Headquarters', 15992), ('Remote', 5255)]

### 5. Could you offer information about the average duration of employment for employees whose tenure has concluded?

    --The average length of employment for terminated employees is calculated by determining the difference between the termination date and hire date for each terminated employee, and then calculating the average of these differences.

In [38]:
%%sql
SELECT ROUND(AVG(DATEDIFF(termdate, hire_date))/365,0) AS avg_length_of_employment
FROM hr
WHERE termdate <> '0000-00-00' AND termdate <= CURDATE() AND age >= 17;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
1 rows affected.


[(Decimal('8'),)]

### 6. How do gender proportions differ across various departments and job titles?
    -- Gender distribution across departments and job titles is achieved by grouping the data by department, job title, and gender, and then counting the occurrences in each group.

In [39]:
%%sql 
SELECT department, gender, COUNT(*) as count
FROM hr
WHERE age >= 17
GROUP BY department, gender
ORDER BY department;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
38 rows affected.


[('Accounting', 'Female', 1462),
 ('Accounting', 'Male', 1640),
 ('Accounting', 'Non-Conforming', 90),
 ('Auditing', 'Female', 23),
 ('Auditing', 'Male', 27),
 ('Business Development', 'Female', 722),
 ('Business Development', 'Male', 798),
 ('Business Development', 'Non-Conforming', 49),
 ('Engineering', 'Female', 2983),
 ('Engineering', 'Male', 3225),
 ('Engineering', 'Non-Conforming', 179),
 ('Human Resources', 'Female', 820),
 ('Human Resources', 'Male', 867),
 ('Human Resources', 'Non-Conforming', 40),
 ('Legal', 'Female', 134),
 ('Legal', 'Male', 158),
 ('Legal', 'Non-Conforming', 7),
 ('Marketing', 'Female', 230),
 ('Marketing', 'Male', 245),
 ('Marketing', 'Non-Conforming', 5),
 ('Product Management', 'Female', 270),
 ('Product Management', 'Male', 338),
 ('Product Management', 'Non-Conforming', 15),
 ('Research and Development', 'Female', 491),
 ('Research and Development', 'Male', 506),
 ('Research and Development', 'Non-Conforming', 35),
 ('Sales', 'Female', 793),
 ('Sales', 'Male', 906),
 ('Sales', 'Non-Conforming', 46),
 ('Services', 'Female', 770),
 ('Services', 'Male', 815),
 ('Services', 'Non-Conforming', 33),
 ('Support', 'Female', 419),
 ('Support', 'Male', 450),
 ('Support', 'Non-Conforming', 34),
 ('Training', 'Female', 759),
 ('Training', 'Male', 819),
 ('Training', 'Non-Conforming', 44)]

### 7. Can you outline the prevalence of different job titles across the entire company?
    -- The distribution of job titles is obtained by grouping the data by job title and counting the occurrences of each job title.

In [41]:
%%sql
SELECT jobtitle, COUNT(*) as count
FROM hr
WHERE age >= 17
GROUP BY jobtitle
ORDER BY jobtitle DESC;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
185 rows affected.


[('Web Developer IV', 66),
 ('Web Developer III', 69),
 ('Web Developer II', 74),
 ('Web Developer I', 95),
 ('Web Designer IV', 6),
 ('Web Designer III', 13),
 ('Web Designer II', 4),
 ('Web Designer I', 34),
 ('VP Sales', 5),
 ('VP Quality Control', 43),
 ('VP Product Management', 39),
 ('VP of Training and Development', 1),
 ('VP Marketing', 42),
 ('VP Accounting', 76),
 ('Training Manager', 177),
 ('Trainer III', 173),
 ('Trainer II', 175),
 ('Trainer I', 171),
 ('Tax Accountant', 294),
 ('Systems Administrator IV', 84),
 ('Systems Administrator III', 81),
 ('Systems Administrator II', 66),
 ('Systems Administrator I', 359),
 ('Support Staff III', 19),
 ('Support Staff II', 58),
 ('Support Staff', 94),
 ('Structural Engineer', 293),
 ('Structural Analysis Engineer', 45),
 ('Statistician IV', 2),
 ('Statistician III', 4),
 ('Statistician II', 11),
 ('Statistician I', 12),
 ('Staff Scientist', 31),
 ('Staff Accountant IV', 101),
 ('Staff Accountant III', 90),
 ('Staff Accountant II', 77),
 ('Staff Accountant I', 427),
 ('Solutions Engineer Manager', 206),
 ('Solutions Engineer', 204),
 ('Software Test Engineer IV', 84),
 ('Software Test Engineer III', 82),
 ('Software Test Engineer II', 69),
 ('Software Test Engineer I', 80),
 ('Software Engineer IV', 88),
 ('Software Engineer III', 97),
 ('Software Engineer II', 80),
 ('Software Engineer I', 378),
 ('Software Consultant', 327),
 ('Service Tech III', 319),
 ('Service Tech II', 315),
 ('Service Tech', 315),
 ('Service Manager', 319),
 ('Service Coordinator', 313),
 ('Senior Trainer', 176),
 ('Senior Sales Associate', 14),
 ('Senior Recruiter', 193),
 ('Senior Quality Engineer', 148),
 ('Senior Financial Analyst', 57),
 ('Senior Editor', 138),
 ('Senior Developer', 346),
 ('Senior Cost Accountant', 301),
 ('Senior Attorney', 37),
 ('Sales Representative', 5),
 ('Sales Associate', 6),
 ('Research Associate', 201),
 ('Research Assistant IV', 6),
 ('Research Assistant III', 18),
 ('Research Assistant II', 718),
 ('Research Assistant I', 507),
 ('Research Assistant', 39),
 ('Relationshiop Manager', 201),
 ('Recruiting Manager', 22),
 ('Recruiter', 222),
 ('Quality Engineer', 61),
 ('Quality Control Specialist', 131),
 ('Project Manager', 329),
 ('Programmer IV', 61),
 ('Programmer III', 78),
 ('Programmer II', 77),
 ('Programmer I', 84),
 ('Programmer Analyst IV', 72),
 ('Programmer Analyst III', 78),
 ('Programmer Analyst II', 89),
 ('Programmer Analyst I', 86),
 ('Product Engineer', 56),
 ('Pre-Sales Consultant', 209),
 ('Payment Adjustment Coordinator', 23),
 ('Paralegal', 63),
 ('Operator', 210),
 ('Office Assistant IV', 1),
 ('Office Assistant II', 1),
 ('Office Assistant I', 7),
 ('Media Manager II', 47),
 ('Media Manager I', 124),
 ('Mechanical Systems Engineer', 28),
 ('Marketing Manager', 1),
 ('Marketing Assistant', 8),
 ('Librarian', 27),
 ('Legal Assistant', 19),
 ('Junior Trainer', 175),
 ('Internal Auditor', 50),
 ('Information Systems Manager', 34),
 ('Human Resources Manager', 58),
 ('Human Resources Assistant IV', 11),
 ('Human Resources Assistant III', 14),
 ('Human Resources Assistant II', 10),
 ('Human Resources Assistant I', 14),
 ('Human Resources Analyst II', 591),
 ('Human Resources Analyst', 393),
 ('HR Manager', 193),
 ('Help Desk Technician', 322),
 ('Help Desk Operator', 88),
 ('Graphic Designer', 12),
 ('General Manager', 32),
 ('Financial Analyst', 65),
 ('Financial Advisor', 27),
 ('Executive Secretary', 2),
 ('Executive Assistant', 38),
 ('Engineer IV', 1),
 ('Engineer III', 5),
 ('Engineer II', 6),
 ('Engineer I', 9),
 ('Electrical Engineer', 29),
 ('Editor', 88),
 ('Director of Sales', 9),
 ('Developer IV', 71),
 ('Developer III', 81),
 ('Developer II', 64),
 ('Developer I', 77),
 ('Desktop Support Technician', 302),
 ('Design Engineer', 50),
 ('Database Administrator IV', 76),
 ('Database Administrator III', 81),
 ('Database Administrator II', 86),
 ('Database Administrator I', 81),
 ('Data Visualization Specialist', 434),
 ('Data Coordiator', 34

### 8. Which department exhibits the highest rate of turnover among its employees?

    -- "Turnover rate" commonly refers to the frequency at which employees exit a company or department, necessitating replacements. It can be computed by dividing the count of departing employees during a specific timeframe by the average number of employees present in the company or department within that corresponding timeframe.

In [42]:
%%sql
SELECT department, COUNT(*) as total_count, 
    SUM(CASE WHEN termdate <= CURDATE() AND termdate <> '0000-00-00' THEN 1 ELSE 0 END) as terminated_count, 
    SUM(CASE WHEN termdate = '0000-00-00' THEN 1 ELSE 0 END) as active_count,
    (SUM(CASE WHEN termdate <= CURDATE() THEN 1 ELSE 0 END) / COUNT(*)) as termination_rate
FROM hr
WHERE age >= 17
GROUP BY department
ORDER BY termination_rate DESC;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
13 rows affected.


[('Marketing', 480, Decimal('43'), Decimal('410'), Decimal('0.9438')),
 ('Training', 1622, Decimal('189'), Decimal('1338'), Decimal('0.9414')),
 ('Human Resources', 1727, Decimal('190'), Decimal('1430'), Decimal('0.9380')),
 ('Services', 1618, Decimal('172'), Decimal('1337'), Decimal('0.9326')),
 ('Accounting', 3192, Decimal('350'), Decimal('2626'), Decimal('0.9323')),
 ('Engineering', 6387, Decimal('690'), Decimal('5259'), Decimal('0.9314')),
 ('Product Management', 623, Decimal('66'), Decimal('512'), Decimal('0.9278')),
 ('Business Development', 1569, Decimal('147'), Decimal('1307'), Decimal('0.9267')),
 ('Sales', 1745, Decimal('189'), Decimal('1426'), Decimal('0.9255')),
 ('Legal', 299, Decimal('39'), Decimal('237'), Decimal('0.9231')),
 ('Auditing', 50, Decimal('8'), Decimal('38'), Decimal('0.9200')),
 ('Research and Development', 1032, Decimal('116'), Decimal('830'), Decimal('0.9167')),
 ('Support', 903, Decimal('92'), Decimal('732'), Decimal('0.9125'))]

### 9.What is the dispersion of employees across various city and state locations?
    -- The distribution of employees across different states is calculated by grouping the data by location state and counting the occurrences in each group.

In [43]:
%%sql
SELECT location_state, COUNT(*) as count
FROM hr
WHERE age >= 17
GROUP BY location_state
ORDER BY count DESC;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
7 rows affected.


[('Ohio', 17252),
 ('Pennsylvania', 1066),
 ('Illinois', 833),
 ('Indiana', 666),
 ('Michigan', 650),
 ('Kentucky', 417),
 ('Wisconsin', 363)]

### Can you determine the historical shifts in the company's employee count over time, considering both hires and terminations?

    -- **This query groups the employees by the year of their hire date and calculates the total number of hires, terminations, and net change (the difference between hires and terminations) for each year. The results are sorted by year in ascending order.**

In [44]:
%%sql
SELECT 
    YEAR(hire_date) AS year, 
    COUNT(*) AS hires, 
    SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END) AS terminations, 
    COUNT(*) - SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END) AS net_change,
    ROUND(((COUNT(*) - SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END)) / COUNT(*) * 100),2) AS net_change_percent
FROM 
    hr
WHERE age >= 17
GROUP BY 
    YEAR(hire_date)
ORDER BY 
    YEAR(hire_date) ASC;

 * mysql+mysqlconnector://root:***@localhost:3306/hr_report
21 rows affected.


[(2000, 211, Decimal('26'), Decimal('185'), Decimal('87.68')),
 (2001, 1082, Decimal('197'), Decimal('885'), Decimal('81.79')),
 (2002, 1012, Decimal('161'), Decimal('851'), Decimal('84.09')),
 (2003, 1088, Decimal('187'), Decimal('901'), Decimal('82.81')),
 (2004, 1087, Decimal('190'), Decimal('897'), Decimal('82.52')),
 (2005, 1038, Decimal('166'), Decimal('872'), Decimal('84.01')),
 (2006, 1069, Decimal('169'), Decimal('900'), Decimal('84.19')),
 (2007, 1058, Decimal('138'), Decimal('920'), Decimal('86.96')),
 (2008, 1061, Decimal('136'), Decimal('925'), Decimal('87.18')),
 (2009, 1094, Decimal('141'), Decimal('953'), Decimal('87.11')),
 (2010, 1050, Decimal('118'), Decimal('932'), Decimal('88.76')),
 (2011, 1057, Decimal('107'), Decimal('950'), Decimal('89.88')),
 (2012, 1059, Decimal('102'), Decimal('957'), Decimal('90.37')),
 (2013, 1042, Decimal('87'), Decimal('955'), Decimal('91.65')),
 (2014, 1014, Decimal('81'), Decimal('933'), Decimal('92.01')),
 (2015, 1011, Decimal('69'), Decimal('942'), Decimal('93.18')),
 (2016, 1076, Decimal('61'), Decimal('1015'), Decimal('94.33')),
 (2017, 1043, Decimal('49'), Decimal('994'), Decimal('95.30')),
 (2018, 1090, Decimal('41'), Decimal('1049'), Decimal('96.24')),
 (2019, 1038, Decimal('44'), Decimal('994'), Decimal('95.76')),
 (2020, 967, Decimal('21'), Decimal('946'), Decimal('97.83'))]

## Summary of Findings

- There are more male employees
- White race is the most dominant while Native Hawaiian and American Indian are the least dominant.
- The youngest employee is 20 years old and the oldest is 57 years old
- 5 age groups were created (17-24, 25-34, 35-44, 45-54, 55-64). A large number of employees were between 25-34 followed by 35-44 while the smallest group was 55-64.
- A large number of employees work at the headquarters versus remotely.
- The average length of employment for terminated employees is around 7 years.
- The gender distribution across departments is fairly balanced but there are generally more male than female employees.
- The Marketing department has the highest turnover rate followed by Training. The least turn over rate are in the Research and development, Support and Legal departments.
- A large number of employees come from the state of Ohio.
- The net change in employees has increased over the years.
- The average tenure for each department is about 8 years with Legal and Auditing having the highest and Services, Sales and Marketing having the lowest.

## Limitations
- Some records had negative ages and these were excluded during querying(967 records). Ages used were 18 years and above.
- Some termdates were far into the future and were not included in the analysis(1599 records). The only term dates used were those less than or equal to the current date.